### ISSUE TITLE
Long scale down time with autoscaling groups

### ISSUE THREAD

**Issue by mrocklin** [pangeo-data/pangeo#322]
Today I spun up the cluster to around 500 nodes for some testing.  Then Joe ran a tutorial with around 10-20 students on the same nodes (I noticed that they were active and so stopped).  The cluster stayed at the heightened state throughout the tutorial, despite the fact that they weren't using nearly this many nodes.  After the tutorial, the cluster wouldn't scale down.  Things cleaned up a bit after I deleted all of the hook-puller daemonsets (@jhamman's recommendation) and then the cluster started to scale down at the rate of one node per minute.  The fact that the cluster hung around in a very active state for the afternoon probably cost us around $300-400.  The fact that it's scaling down very slowly right now will cost us another $300 or so.  This is fine, we're learning, but we do probably need to learn from this.

I think that there are two questions:

1.  What are the hook-pullers, why were there 7 daemonsets for them, and why is it that they seem to stop a smooth scaling down of the cluster?  (also is this the case or was it just a coincidence that things started scaling down more smoothly after they were deleted)  cc @yuvipanda
2.  How can we get Google's autoscaling node pools to scale down more quickly.  This is important for the burst-y computational workloads that we expect.  cc @lila

**Comment by yuvipanda** [comment]
I think the hook-pullers are a co-incidence. What version of z2jh are you running? @consideRatio has done a lot of work around this.

GKE's autoscaler is an installation of https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler that comes by default and has parameters you can not change. You can deploy a copy of that yourself instead of the default autoscaler & tune it to your needs.

**Comment by consideRatio** [comment]
Thanks for bringing this to my attention @yuvipanda!

@mrocklin yikes!

1. They are meant to prepare the cluster by pulling images to the nodes before a upgrade is made, but due to some helm issues they don't delete properly. This is fixed in 0.7-dev releases of the [zero-to-jupyterhub-k8s](https://github.com/jupyterhub/zero-to-jupyterhub-k8s) helm chart. I'm actually working right now with optimizing the autoscaling in robustness and cost.

I highly recommend that you disable the hook image-puller in `config.yaml` that you provide helm under `prePuller.hook.enabled: false`. The continuous-image-puller makes a bit of sense for your cluster though, it will start pulling images to new nodes when they are added. Nothing vital.

2. You can configure that by editing a ConfigMap located in the kube-system namespace I think... I remember it being set to 10 minutes as a default. Hmmm... I was mistaken... I figure you need to deploy it on your own as if you don't it will be unconfigurable as far as i know on GKE as it runs on the non-accessible kubernetes master. Please tell me if you figure out how to do it though, it might look for a configmap or so...

**Comment by mrocklin** [comment]
This is happening again now.  @martindurant if you have time next week can I ask you to look into this.  This seems fairly high priority if we want to preserve our allocation.

**Comment by jhamman** [comment]
I'm confused by this: 

```
kubectl get pods --namespace pangeo
NAME                      READY     STATUS              RESTARTS   AGE
hub-6ff5d74d5b-mf8p5      1/1       Running             1          7h
jupyter-andersy005        1/1       Running             0          3h
jupyter-jiaweizhuang      1/1       Running             0          3h
jupyter-rsignell-2dusgs   0/1       ContainerCreating   0          2m
proxy-64f95bb784-xhhl4    1/1       Running             0          7h

```

yet the preimtible pool is sitting at 499 nodes. So what are the 496 pods doing?

**Comment by martindurant** [comment]
Running a resize command now, it's taking very long, perhaps as long as Matt was trying to do in the web UI.

**Comment by martindurant** [comment]
Note that we have persistent 1537 disks

**Comment by tjcrone** [comment]
We also have a large number of "succeeded" pods.
```
kubectl describe pods -n pangeo | grep -E "^Name:|^Node:|^Status:"
```
I don't really understand what a "succeeded" pod is, but they tend to never go away unless they are deleted explicitly, and are not listed by get pods. I delete them on occasion from my cluster because they can pile up. I don't think they prevent a scale down, so I'm not sure what is going on there. Maybe they do slow down the autoscaler? I don't know.

**Comment by martindurant** [comment]
Error on node creation:
```
  Warning  FailedToCreateRoute      1h                 route_controller                                       Could not create route 23cecfae-853e-11e8-b8a7-42010a80011f 10.21.40.0/24 for node gke-pangeo-preemptible-pool-14c39f44-6f1c after 1.951932097s: googleapi: Error 403: Quota 'ROUTES' exceeded. Limit: 300.0 globally., quotaExceeded
```

**Comment by martindurant** [comment]
I wouldn't be sure that succeeded pods get cleared up: they should have (the equivalent of) the docker `--rm` flag applied on creation.

**Comment by martindurant** [comment]
I also notice that the default-pool is in us-central1-a, but the rest of the instances are in us-central1-b. Sorry for this list of findings, all possibly unrelated.

**Comment by tjcrone** [comment]
Maybe they get cleared up eventually. They do hang around for a very long time on my cluster. There are currently 646 succeeded pods on the cluster, all of which were started today. If they do prevent a scale down, could be what is slowing things down.

**Comment by martindurant** [comment]
Instance numbers have started to go down. I issues a load of kubectl delete node, and then gcloud delete instance commands, as well as resize ones that seem to be pending. I don't know whether any have had any effect, or whether the system is just catching up with work now.

**Comment by consideRatio** [comment]
Hey! I read this through and the following things crossed my mind.

## Cluster autoscaler inspection
It is hard to debug whats going on in the cluster autoscaler, but this is what you can find out without access to the kubernetes master:
- https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-events-are-emitted-by-ca

## Clearing of orphaned pre-pullers
To clear a lot of junk related to orphaned pre-pullers (see https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/638)
```
kubectl delete ds,sa,clusterrole,clusterrolebinding,job --selector hub.jupyter.org/deletable=true
```

( copied from https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/638 )

## Blocking pods
I don't know about the pods not being deleted left in the Succeeded phase, but if you are using 500 nodes, you will have a kube-dns pod for every 16th node. That pod will block downscaling. You can configure this to be one for every 32th node etc within a config file you will find in the kube-system namespace.

## Future autoscaling improvements
I'm working on autoscaling in a robust manner currently, a custom scheduler is introduced to help out packing pods together instead of spreading out which isn't very clever in a autoscaling environment as you the autoscaler will never be able to scale down if pods keep spreading out.

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/758

**Comment by mrocklin** [comment]
I suggest that we also try to use fewer/bigger nodes at some point soon.

**Comment by mrocklin** [comment]
Also cc @lila who might be able to upstream some of this experience to
engineers at Google.  The slowness of the default scale down behavior of
adaptive node pools (reduce the size by one node per minute) is quite
expensive for groups that have very burst-y usage patterns, such as is
common in interactive data analysis.  This has been an expensive lesson for
us.

**Comment by martindurant** [comment]
Continuing to see the 500 server-error upon trying to login, even now the cluster has stabilised. 
Aside from the node creation error I posted above ( `googleapi: Error 403: Quota 'ROUTES' exceeded `), I haven't seen anything amiss, the pod created and jupyter is waiting, but the hub doesn't find it. Note that I don't have access to the quotas part of the compute engine UI, so I don't know if we've run out of anything else.

**Comment by yuvipanda** [comment]
I think @consideRatio's work on extra schedulers has been merged upstream now, and lessons from that would probably also be useful for the dask worker pods.

**Comment by consideRatio** [comment]
Merged! And the awesome graphs on mybinder.org provided [some fun insights](https://grafana.mybinder.org/d/nDQPwi7mk/node-activity?orgId=1&from=1536218397087&to=1536243497087&panelId=26&fullscreen&refresh=1m)

![image](https://user-images.githubusercontent.com/3837114/45192604-91023d80-b249-11e8-8d84-872be7136f94.png)

**Comment by consideRatio** [comment]
See https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/841 to follow progress on additional improvements, such as overprovisioning pods to allow scale up in advance. From this graph it seems like it would be suitable to have about 10 placeholder pods creating headroom. I conclude that from looking at the points where a fourth node is added back for short durations and seeing there were a jump from 0 to 5 or so pods.

**Comment by yuvipanda** [comment]
@consideRatio Best way to learn about routes is to learn about the Kubernetes Pod Network model. This is a somewhat complex topic, but there's lots of info out there on this. Googling gave me https://medium.com/@ApsOps/an-illustrated-guide-to-kubernetes-networking-part-1-d1ede3322727 for example.

**Comment by yuvipanda** [comment]
I got a response back from them basically asking to use https://cloud.google.com/kubernetes-engine/docs/how-to/alias-ips. I'll open a new issue.

**Comment by jhamman** [comment]
@dsludwig - this is another issue that could use your help. Both for pangeo jupyterhubs and for binder.

**Comment by dsludwig** [comment]
This seems like it could be relevant to the dask-worker pods:

> What types of pods can prevent CA from removing a node?
> ...
> Pods that are not backed by a controller object (so not created by deployment, replica set, job, stateful set etc). *

From https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node

**Comment by dsludwig** [comment]
If we could configure the autoscaler, there's some things we could do to tune how the cluster scales: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#i-have-a-couple-of-nodes-with-low-utilization-but-they-are-not-scaled-down-why

> * the node group already has the minimum size,
> * node has the scale-down disabled annotation (see How can I prevent Cluster Autoscaler from scaling down a particular node?)
> * node was unneeded for less than 10 minutes (configurable by --scale-down-unneeded-time flag),
> * there was a scale-up in the last 10 min (configurable by --scale-down-delay-after-add flag),
> * there was a failed scale-down for this group in the last 3 minutes (configurable by --scale-down-delay-after-failure flag),
> * there was a failed attempt to remove this particular node, in which case Cluster Autoscaler will wait for extra 5 minutes before considering it for removal again,
> * using large custom value for --scale-down-delay-after-delete or --scan-interval, which delays CA action.

However, it's not possible to configure the cluster autoscaler when creating node pools on GKE using the `--autoscaling` flag: https://github.com/kubernetes/autoscaler/issues/966

**Comment by consideRatio** [comment]
@dsludwig In the long run, one may be allowed from a GKE cluster to configure the autoscaler by having access to a custom resource that it will look for. I would not sit around and wait for it though, they removed such functionality since it wasn't utilized enough and added some complexity if i don't remember incorrectly.

I'm not sure how your pods are scheduled on the cluster, but have you made sure that nodes are getting empty quickly? You would probably want a custom scheduler as is now available in the zero-to-jupyterhub-k8s master. You can also utilize the user-placeholder statefulset to scale up in advance and the k8s 1.11+ pod-priority feature allowing you to evict those pods in favor of a non-placeholder pod if it would failing to schedule without waiting for a scale up.

If there is a pattern of a group of pods going to be deleted at the same time, you should make them have affinity, so that a large portion or a full node may go empty when they all complete. But if you are speaking of clusters of 500 nodes, inter pod affinitys performance may be quite bad... But, in k8s 1.12 that performance is improved 100x at least.

Just a flow of ideas here... Hope something may be of relevance, if you could present a clear usage pattern of how pods are scheduled and going over time, you or someone else may be able to better say what would be a good way to maximize resource utilization and reduce costs.

**Comment by consideRatio** [comment]
@mrocklin btw there will only be at most one hook puller in the current z2jh master, and they will go away whenever a helm upgrade successfully completes.

**Comment by dsludwig** [comment]
Hmm... it looks like we've got some kube-system pods for critical services (tiller, heapster, metrics-server) migrating to expensive/user pool nodes, and that's preventing scale-down. We may need to taint the user pools to ensure they don't get there in the first place.

Alternatively, we could add pod disruption budgets for those system services, to allow them to be unavailable while scaling down (https://github.com/kubernetes/autoscaler/issues/64).

**Comment by consideRatio** [comment]
yepp i recommend having a few nodes for kube dns and various other pods that scale up with node count, and then a node pool for the huge number of pods that you burst living with tolerations on tainted nodes.

support for setting tolerations for the jh user pods is around in current z2jh master btw

**Comment by MaciekPytel** [comment]
Hi, Cluster Autoscaler developer here. 

In general autoscaler can remove multiple nodes concurrently if those nodes are completely empty. Draining nodes in parallel is not supported by autoscaler regardless of what set of flags you use. So if some other pods get scheduled on your nodes while your batch is running it will likely take a significant amount of time to scale-down. 

I would suggest using a dedicated node pool for your batch. You can do this by creating a node pool with a specific taint and label. Than add a toleration and nodeSelector to your batch workload. This ensures that your batch will always run on dedicated nodes and no other pod will be able to run on those nodes. Once the batch finishes the nodes should be completely empty allowing autoscaler to remove them quickly. This is basically what @dsludwig suggested.

**Comment by glapark** [comment]
> Hi, Cluster Autoscaler developer here.
> 
> In general autoscaler can remove multiple nodes concurrently if those nodes are completely empty. Draining nodes in parallel is not supported by autoscaler regardless of what set of flags you use. So if some other pods get scheduled on your nodes while your batch is running it will likely take a significant amount of time to scale-down.
> 
> I would suggest using a dedicated node pool for your batch. You can do this by creating a node pool with a specific taint and label. Than add a toleration and nodeSelector to your batch workload. This ensures that your batch will always run on dedicated nodes and no other pod will be able to run on those nodes. Once the batch finishes the nodes should be completely empty allowing autoscaler to remove them quickly. This is basically what @dsludwig suggested.

Hello, if a node is completely empty, does 

> Hi, Cluster Autoscaler developer here.
> 
> In general autoscaler can remove multiple nodes concurrently if those nodes are completely empty. Draining nodes in parallel is not supported by autoscaler regardless of what set of flags you use. So if some other pods get scheduled on your nodes while your batch is running it will likely take a significant amount of time to scale-down.
> 
> I would suggest using a dedicated node pool for your batch. You can do this by creating a node pool with a specific taint and label. Than add a toleration and nodeSelector to your batch workload. This ensures that your batch will always run on dedicated nodes and no other pod will be able to run on those nodes. Once the batch finishes the nodes should be completely empty allowing autoscaler to remove them quickly. This is basically what @dsludwig suggested.

Hello. If a node is completely empty, is it still removed after the period specified by --scale-down-unneeded-time (default 10 minutes)? It seems like --scale-down-unneeded-time is not configurable on GKE, and I wonder if emptying nodes helps the autoscaler to quickly remove them at all.

Thanks,

**Comment by MaciekPytel** [comment]
Currently CA starts deleting nodes after 10 minutes regardless if they're empty or not. The difference between empty / non-empty nodes is how quickly it can remove them once it starts. Empty nodes are deleted in parallel, draining nodes is done one-by-one. The throughput (#deleted nodes / minute) can be >10x higher for empty nodes.
